<a href="https://colab.research.google.com/github/Vitela-debug/Colab_S1_Clase-de-progralaci-n_273522_TVB/blob/main/Actividad_Calificada_07_Datos_Temporales_273522.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo: Tania Vitela Barraza** | |
| **Matrícula: 273522** | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [33]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [34]:
# Escribe tu código aquí
#1
print('Tipo de dato pickip antes de convesión', taxis['pickup'].dtype) #.dtype para ver el tipo de dato antes de la convesión
print('Tipo de dato dropoff antes de convesión', taxis['dropoff'].dtype)
Faltantes_pickup1 = taxis['pickup'].isna().sum() # .isna().sum()  para ver faltantes
Faltantes_dropoff1 = taxis['dropoff'].isna().sum()
print('Faltantes pickup antes de convesión', Faltantes_pickup1)
print('Faltantes dropoff antes de convesión', Faltantes_dropoff1)
#2
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S') #Asignar el nuevo formato especificando el tipo de date
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')

#3
Faltantes_pickup2 = taxis['pickup'].isna().sum()
Faltantes_dropoff2 = taxis['dropoff'].isna().sum()
print('Faltantes pickup después de convesión', Faltantes_pickup2)
print('Faltantes dropoff después de convesión', Faltantes_dropoff2)
print('Tipo de dato pickip después de convesión', taxis['pickup'].dtype) #.dtype para ver el tipo de dato después de la convesión
print('Tipo de dato dropoff después de convesión', taxis['dropoff'].dtype)

Tipo de dato pickip antes de convesión object
Tipo de dato dropoff antes de convesión object
Faltantes pickup antes de convesión 0
Faltantes dropoff antes de convesión 0
Faltantes pickup después de convesión 0
Faltantes dropoff después de convesión 0
Tipo de dato pickip después de convesión datetime64[ns]
Tipo de dato dropoff después de convesión datetime64[ns]


---
---
* El código del ejercicio 1.1 primero utiliza `.dtype` para revisar qué tipo de dato tienen las columnas pickup y dropoff, y `.isna().sum()` para contar cuántos valores faltantes tiene cada una antes de hacer cambios. Después utiliza `pd.to_datetime()` para convertir ambas columnas de texto a un formato de fecha y hora, especificando con format cómo están organizados los datos (año-mes-día hora:minuto:segundo). Para finalizar usa `.isna().sum() `para comprobar si cambiaron los valores faltantes y `.dtype` para verificar que las columnas ahora tengan el tipo de dato de fecha (datetime). En general el código revisa el tipo de dato y los valores faltantes, convierte las columnas a formato de fecha y hora y después comprueba que la conversión se haya realizado correctamente.
---
---


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [35]:
# Escribe tu código aquí

#1. Ver fecha mínima con .min() y  máxima con max()
print('Fecha mínima de pickup', taxis['pickup'].min())
print('Fecha máxima de pickup', taxis['pickup'].max())
print('Fecha mínima de dropoff', taxis['dropoff'].min())
print('Fecha máxima de dropoff', taxis['dropoff'].max())


#2
Pickup_not_march =taxis[~taxis['pickup'].between('2019-03-01', '2019-03-31')]  #.between para ver rangos ~ para que sean los que estan fuera de esos rangos

print('\n Viajes con pickup fuera de marzo de 2019:', len(Pickup_not_march)) #se cueentan las filas del rango anterior
print(Pickup_not_march)

#3
Dropoff_not_march = taxis[~taxis['dropoff'].between('2019-03-01', '2019-03-31')] #.between para ver rangos ~ para que sean los que estan fuera de esos rangos
Both_not_march = taxis[(~taxis['pickup'].between('2019-03-01', '2019-03-31')) & (~taxis['dropoff'].between('2019-03-01', '2019-03-31'))] #se juntan ambas condiciones con &
print('\n Viajes con pickup y dropoff fuera de marzo de 2019:', len(Both_not_march)) #Se cuentan las que cumplen ambas condiciones


Fecha mínima de pickup 2019-02-28 23:29:03
Fecha máxima de pickup 2019-03-31 23:43:45
Fecha mínima de dropoff 2019-02-28 23:32:35
Fecha máxima de dropoff 2019-04-01 00:13:58

 Viajes con pickup fuera de marzo de 2019: 188
                  pickup             dropoff  passengers  distance  fare  \
69   2019-03-31 15:43:37 2019-03-31 15:53:39           1      0.55   7.5   
103  2019-03-31 11:50:36 2019-03-31 11:54:00           4      0.43   4.0   
125  2019-03-31 17:17:53 2019-03-31 17:29:00           1      2.07  10.0   
141  2019-03-31 08:55:25 2019-03-31 09:00:23           1      1.10   6.0   
164  2019-03-31 15:21:11 2019-03-31 15:37:28           2      1.80  11.0   
...                  ...                 ...         ...       ...   ...   
6287 2019-03-31 14:45:30 2019-03-31 14:49:38           1      0.84   5.0   
6323 2019-03-31 09:35:52 2019-03-31 09:44:22           1      1.67   8.5   
6375 2019-03-31 18:50:27 2019-03-31 18:53:53           1      0.59   4.5   
6428 2019-03-31 09

---
---
* El código del ejercicio 1.2 sirve para revisar el rango de fechas de los viajes y detectar cuáles viajes están fuera de marzo de 2019. Primero, `.min()` busca la fecha más antigua y `.max()` la fecha más reciente de pickup y dropoff, para conocer el periodo que abarcan los datos. Después, `.between()` verifica si las fechas están entre el 1 y el 31 de marzo de 2019; el `~` invierte el resultado, por lo que selecciona las fechas que están fuera de marzo. Finalmente, `len()` cuenta cuántos viajes cumplen esas condiciones. En la última parte, se usan dos condiciones con `&` para encontrar únicamente los viajes en los que tanto pickup como dropoff están fuera de marzo de 2019.
---
---

### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [36]:
# Escribe tu código aquí

viajes_10_marzo = taxis[taxis['pickup'].dt.date == pd.Timestamp('2019-03-10').date()] #Filtra del pickup y los que son del 10 de marzo
viajes_por_hr = viajes_10_marzo['pickup'].dt.hour.value_counts().sort_index() #.dt.hour extrae la hora del viaje, .value_counts() cuenta los viajes por hora y .sort_index() las ordena
print(viajes_por_hr)

if 2 in viajes_por_hr:
  print('Hay viajes a las 2:00-2:59') #Comprobamos con un if que a las 2 no se tienen registrados viajes
else:
  print('No hay viajes a las 2:00-2:59')

registros_ZONE = taxis['pickup'].dt.tz_localize('America/New_York',nonexistent='NaT') #.dt.tz_localize()  asigna una zona horaria a las fechas y nonexistent='NaT' cuando no existe la hora lo asigna
print('Valores con NaT en pickup', registros_ZONE.isna().sum()) #Cuenta los NaT




pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64
No hay viajes a las 2:00-2:59
Valores con NaT en pickup 0


---
---
* El código del ejercicio 1.3 primero filtra los viajes realizados el 10 de marzo de 2019 usando la columna pickup. Con `.dt.hour` extrae la hora de cada viaje y `.value_counts()` cuenta cuántos viajes hubo en cada hora, `.sort_index()` ordena las horas. Luego se utiliza un `if` para comprobar si existe la hora 2, indicando si hubo o no viajes entre las 2:00 y 2:59. Finalmente, .`dt.tz_localize()` asigna la zona horaria America/New_York a pickup y, con `nonexistent='NaT'`, convierte en NaT las horas que no existen, `.isna().sum() `cuenta cuántos registros quedaron como NaT.
---
---

### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.




**Respuesta 1.4:**

* Se encontró un problema en el rango de fechas, porque el dataset debería contener únicamente viajes de marzo de 2019, pero existen registros fuera de ese periodo. La fecha mínima de pickup es 28 de febrero de 2019 a las 23:29:03 y la fecha máxima de dropoff es 1 de abril de 2019 a las 00:13:58. Esto indica que algunos viajes cruzan los límites de marzo (188 datos en total), por lo que habría que definir un criterio para decidir si se conservan o eliminan.

* El 10 de marzo de 2019 no se encontraron viajes entre las 02:00 y 02:59; el conteo pasa de 7 viajes a la 1:00 a 6 viajes a las 3:00. Además, al aplicar la zona horaria America/New_York con nonexistent='NaT', se obtuvieron 0 valores NaT. Esto indica que los datos son compatibles con el cambio de horario de Nueva York, ya que no contienen registros durante la hora que no existió ese día. Sin embargo, estos resultados por sí solos no permiten asegurar que los datos estén originalmente registrados en la zona horaria de Nueva York; solamente muestran que no hay conflictos al interpretarlos con esa zona horaria.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [37]:
# Escribe tu código aquí

#1
taxis['duracion'] = taxis['dropoff']-taxis['pickup'] #hacemos la resta de 2 columnas para obtener la columna de dudacion

print(taxis['duracion'].head(5))

#2
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds()/60 #La diferencia la convertimos en segundos y la dividimos en 60 para tener los minutos

print(taxis['duracion_min'].head(5))

#3
taxis['duracion_min'].describe() #Sacamos el resumen estadistico con describe()

0   0 days 00:06:15
1   0 days 00:07:05
2   0 days 00:07:24
3   0 days 00:25:52
4   0 days 00:09:32
Name: duracion, dtype: timedelta64[ns]
0     6.250000
1     7.083333
2     7.400000
3    25.866667
4     9.533333
Name: duracion_min, dtype: float64


,duracion_min
count,6433.000000
mean,14.349617
std,11.643892
min,0.000000
25%,6.500000
50%,10.900000
75%,18.516667
max,107.666667


---
---
* El código del ejercicio 2.1 primero crea una nueva columna llamada duracion restando la fecha y hora de pickup a la de dropoff, para obtener cuánto duró cada viaje. Después, con `.dt.total_seconds()`, convierte esa duración a segundos y la divide entre 60 para obtener la duración de cada viaje en minutos, guardándola en duracion_min. Y utiliza `.describe()` para obtener un resumen estadístico de la duración de los viajes, mostrando datos como la cantidad de registros (count), promedio (mean), desviación estándar (std), valor mínimo (min), cuartiles (25%, 50%, 75%) y valor máximo (max).
---
---

### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [38]:
# Escribe tu código aquí

D_negativa = taxis[taxis['duracion_min'] < 0] #Viajes con duracion negativa
D_cero = taxis[taxis['duracion_min'] == 0] #Viajes con duracion cero
D_cero_min = taxis[taxis['duracion_min'] < 1] #Viajes con duracion menor a 1 minuto
D_mayor_3 = taxis[taxis['duracion_min'] > 180] #Viajes con duracion mayor a 3 horas (60*3=180)

print('Viajes con duracion negativa:', len(D_negativa)) #Cotamos cada una de las duraciones
print('Viajes con duracion cero:', len(D_cero))
print('Viajes con duracion menor a 1 minuto:', len(D_cero_min))
print('Viajes con duracion mayor a 3 horas:', len(D_mayor_3))

#Mostramos los viajes de duración cero
print('\nViajes con duracion cero:')
print(D_cero[['pickup', 'dropoff', 'distance', 'fare', 'total']])


Viajes con duracion negativa: 0
Viajes con duracion cero: 6
Viajes con duracion menor a 1 minuto: 72
Viajes con duracion mayor a 3 horas: 0

Viajes con duracion cero:
                  pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


---
---
* El código del ejercicio 2.2 revisa la columna duracion_min para detectar duraciones fuera de lo normal. Con` taxis[taxis['duracion_min'] < 0] `identifica viajes con duración negativa, con `== 0` los viajes de duración cero, con `< 1` los menores a un minuto y con `> 180` los que duran más de 3 horas. Después, usando `len(...)`, cuenta cuántos viajes hay en cada caso y, finalmente, con `D_cero[['pickup', 'dropoff', 'distance', 'fare', 'total']]` muestra algunos datos de los viajes que tienen duración cero para poder revisarlos.
---
---

### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [39]:
# Escribe tu código aquí
import numpy as np
#1
velocidad_mph = taxis['distance'] / (taxis['duracion_min'] / 60) #Calculamos la velocidad
taxis['velocidad_mph'] = velocidad_mph.where(taxis['duracion_min'] > 0, np.nan) #se restringe con where y se condicionan a NaN los que son menores a 0
print('Conversiones a NaN:',taxis['velocidad_mph'].isna().sum())

#2
viajes_velocidad_60 = taxis[taxis['velocidad_mph'] > 60] #Viajes con velocidad mayor a 60 con condicional >
print('Viajes con velocidad mayor a 60 mph:', len(viajes_velocidad_60)) #contar los viajes con velocidades mayores a 60

#3
taxis_odenado = taxis.sort_values(by='velocidad_mph', ascending=False) #Ordenar los viajes por velocidad de mayor a menor

print('\n5 viajes con mayor velocidad:')
print(taxis_odenado[['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']].head(5)) #mostar ciertas columnas de los datos ordenados por velocidades



Conversiones a NaN: 6
Viajes con velocidad mayor a 60 mph: 11

5 viajes con mayor velocidad:
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


---
---
* El código del ejercicio 2.3 calcula la velocidad de cada viaje en mph, usando la distancia (distance) y la duración (duracion_min). Con .where() controla los casos donde la duración no es válida. Después, con filtros ([] y >) saco los viajes que superan las 60 mph y con len() cuento cuántos hay. Finalmente, uso .sort_values() para ordenar los viajes de mayor a menor velocidad y .head(5) para mostrar los 5 más rápidos, seleccionando solo algunas columnas.
---
---


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

In [40]:
#Mostramos los viajes de duración menor a 1 min
print('\nViajes con duracion menor a un min:')
print(D_cero_min[[ 'distance', 'fare']].head(10))


Viajes con duracion menor a un min:
     distance   fare
42       0.00   80.0
88       0.07    2.5
120      0.00    2.5
400      0.90    2.5
458      0.02    2.5
491      1.60    2.5
622      0.00  120.0
624      0.31    3.0
644      0.30    3.0
670      0.00    2.5


**Respuesta 2.4:**

a) No, no todos los viajes de menos de 1 minuto son errores. Al observar la distancia y la tarifa, algunos parecen viajes cortos que pueden ser reales, por ejemplo los de 0.07, 0.30 o 0.90 millas con tarifas de $$2.50 a $3.00. Sin embargo, hay casos más sospechosos, como viajes de 0 minutos y 0 millas con tarifas de $80 o $120, que podrían ser errores o registros anómalos.

b) Usaría como umbrales una duración mayor a 0 minutos, porque con duración cero no se puede calcular correctamente la velocidad; una velocidad máxima de 60 mph, porque los valores superiores generan velocidades exageradas como 4834 mph; y un periodo de pickup dentro de marzo de 2019, para mantener todos los datos en el mismo periodo. No eliminaría automáticamente los viajes menores a 1 minuto, ya que algunos tienen distancias y tarifas que pueden ser razonables.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [41]:
# Escribe tu código aquí

taxis['hora'] = taxis['pickup'].dt.hour #Extraemos la hora de pickup
taxis['dia_num'] = taxis['pickup'].dt.dayofweek #Extraemos el día de la semana de pickup
taxis['dia_nombre'] = taxis['pickup'].dt.day_name() #Extraemos el nombre del día de pickup
taxis['fin_semana'] = taxis['dia_num'].isin([5, 6]) #Extraemos si es fin de semana, sábado (5) y domingo (6)
taxis['fecha'] = taxis['pickup'].dt.date #Extraemos la fecha de pickup

taxis['fecha_dro'] = taxis['dropoff'].dt.date
taxis['cruza_medianoche'] = taxis['fecha'] !=  taxis['dropoff'].dt.date #Extraemos si cruza la medianoche


print(taxis[['hora', 'dia_num', 'dia_nombre', 'fin_semana', 'fecha', 'cruza_medianoche', 'fecha_dro']].head(10)) #mostrar las primeras 10 filas


print('\n Viajes que cruzan la medianoche:', taxis['cruza_medianoche'].sum()) #Contar viajes con criza a medianoche

   hora  dia_num dia_nombre  fin_semana       fecha  cruza_medianoche  \
0    20        5   Saturday        True  2019-03-23             False   
1    16        0     Monday       False  2019-03-04             False   
2    17        2  Wednesday       False  2019-03-27             False   
3     1        6     Sunday        True  2019-03-10             False   
4    13        5   Saturday        True  2019-03-30             False   
5    10        0     Monday       False  2019-03-11             False   
6    21        1    Tuesday       False  2019-03-26             False   
7    12        4     Friday       False  2019-03-22             False   
8    11        5   Saturday        True  2019-03-23             False   
9    16        4     Friday       False  2019-03-08             False   

    fecha_dro  
0  2019-03-23  
1  2019-03-04  
2  2019-03-27  
3  2019-03-10  
4  2019-03-30  
5  2019-03-11  
6  2019-03-26  
7  2019-03-22  
8  2019-03-23  
9  2019-03-08  

 Viajes que cruzan 

---
---
* En el código del ejercicio 3.1 se utilizó `taxis['pickup'].dt `para extraer información de la fecha y hora del pickup. Con `.dt.hour` obtengo la hora, con .`dt.dayofweek` saco el número del día de la semana y con .dt.day_name() obtengo su nombre. Después uso `.isin([5, 6]) `para identificar si el viaje fue en fin de semana, considerando 5 como sábado y 6 como domingo. Con .`dt.date `extraigo solamente la fecha del pickup y del dropoff. Finalmente, con `!=` comparo ambas fechas para saber si el viaje cruzó la medianoche, y con `.head(10)` muestro las primeras 10 filas. También utilizo `.sum()` sobre cruza_medianoche para contar cuántos viajes terminaron en un día diferente al que comenzaron.
---
---

### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [42]:
# Escribe tu código aquí

Grupos = taxis.groupby('hora').size() #Se calcula el 'size' de cada hora
Grupos_mediana = taxis.groupby('hora')['duracion_min'].median().round(1) #Se calcula la mediana de duración por hora
Grupos_hora = pd.concat([Grupos, Grupos_mediana], axis=1) #Juntamos las dos columnas
Grupos_hora.columns = ['cantidad_viajes', 'mediana_duración'] #Asignamos nomvre a las columnas
print(Grupos_hora)


      cantidad_viajes  mediana_duración
hora                                   
0                 205              10.7
1                 110               9.6
2                 101               9.2
3                  67               9.6
4                  57              10.5
5                  51               8.7
6                 142               7.8
7                 221               9.6
8                 313              11.3
9                 319              11.6
10                325              11.9
11                295              12.0
12                334              12.8
13                316              10.6
14                360              11.8
15                330              12.1
16                336              11.6
17                388              12.0
18                417              11.1
19                406              10.6
20                368               9.3
21                355              11.1
22                321               9.8


---
---
*  El código del ejercicio 3.2 agrupa los viajes por hora para analizar cuántos viajes hay y cuánto duran normalmente. Con `taxis.groupby('hora').size() `cuento los viajes de cada hora, mientras que `taxis.groupby('hora')['duracion_min'].median().round(1)` calcula la mediana de la duración y la redondea a un decimal. Después, `pd.concat([Grupos, Grupos_mediana], axis=1)` junta ambos resultados en una sola tabla para poder compararlos por hora.
---
---


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [43]:
# Escribe tu código aquí
grupos_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg( #se agrupa por dia y nombre y .agg() para aplicar diferentes operaciones a cada columna
    viajes=('pickup', 'count'), #Cuenta los viajes por dia
    mediana_duracion=('duracion_min', 'median'), #La mediana por duracion
    mediana_fare=('fare', 'median') #la mediana por tarifa
)

fechas_dia = taxis.groupby(['dia_num', 'dia_nombre'])['fecha'].nunique() #cuántas fechas diferentes existen para cada día

grupos_dia['promedio_viajes_fecha'] = grupos_dia['viajes'] / fechas_dia #calcula el promedio de viajes por fecha

grupos_dia = grupos_dia.round(1)

grupos_dia

,,viajes,mediana_duracion,mediana_fare,promedio_viajes_fecha
dia_num,dia_nombre,,,,
0,Monday,708,10.5,9.5,177.0
1,Tuesday,825,11.2,9.5,206.2
2,Wednesday,966,11.3,9.5,241.5
3,Thursday,905,11.8,9.5,181.0
4,Friday,1115,11.2,9.5,223.0
5,Saturday,1046,10.5,9.0,209.2
6,Sunday,868,10.0,9.0,173.6


---
---
* El código del ejercicio 3.3 agrupo los viajes por dia_num y dia_nombre usando `groupby()` para tener los días de la semana ordenados de lunes a domingo. Con .`agg() `calculo el número de viajes contando los registros de pickup, además de la `mediana_duracion` y la` mediana_fare`. Después, con` nunique()` cuento cuántas fecha diferentes hay para cada día de la semana, ya que marzo no tiene la misma cantidad de cada día. Luego divido viajes entre esas fechas para obtener `promedio_viajes_fecha`
---
---


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) Al comparar los días, observo que viernes es el día con más viajes, con 1115, mientras que lunes tiene menos, con 708. La duración mediana es bastante parecida durante toda la semana, aunque jueves tiene la mayor duración mediana (11.8 minutos) y domingo la menor (10.0 minutos). Por lo tanto, la hora con más viajes y la hora con mayor duración mediana no necesariamente coinciden. Esto puede explicarse porque tener muchos viajes en una hora no significa que esos viajes duren más; la duración puede depender del tráfico, la distancia de los recorridos o las condiciones de cada momento.

b) Fue necesario calcular el promedio de viajes por fecha porque marzo de 2019 no tiene la misma cantidad de lunes, martes, miércoles, etc. Si solo comparáramos el total de viajes, un día de la semana que aparece más veces podría parecer que tiene más viajes simplemente porque tuvo más fechas disponibles. Al dividir los viajes entre el número de fechas distintas usando `nunique()`, podemos comparar cuántos viajes se realizan en promedio en cada día, haciendo la comparación más justa.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [44]:
# Escribe tu código aquí

# Condiciones para excluir viajes
cond_duracion = taxis['duracion_min'] <= 0 #menores o iguales que cero
cond_velocidad = taxis['velocidad_mph'] > 60 #velocidades que superar las 60 mph
cond_fecha = ~taxis['pickup'].between('2019-03-01', '2019-03-31 23:59:59') #Excluir los datos fuera de marzo 2019

# Contar viajes excluidos por cada condición
print("Viajes excluidos por duración:", cond_duracion.sum()) #contar los datos excluido por duracion
print("Viajes excluidos por velocidad:", cond_velocidad.sum()) #contar los datos excluido por velocidad
print("Viajes excluidos por fecha:", cond_fecha.sum()) #contar los datos excluido por fecha

# Crear el dataset limpio
taxis_limpio = taxis[ ~(cond_duracion | cond_velocidad | cond_fecha)].copy() #para quitar los datos que no se quieren

# Contar el total de viajes excluidos
print("Total de viajes excluidos:", len(taxis) - len(taxis_limpio)) #Total de excluidos

# Verificar que taxis_limpio cumple las condiciones
print("\nVerificación:")
print("Duraciones <= 0:", (taxis_limpio['duracion_min'] <= 0).sum()) #Contar duraciones menores de 0 en taxis_limpio
print("Velocidades > 60 mph:", (taxis_limpio['velocidad_mph'] > 60).sum()) #contar velocidades mayores de 60 en taxis_limpio
print("Pickups fuera de marzo 2019:", (~taxis_limpio['pickup'].between('2019-03-01', '2019-03-31 23:59:59')).sum()) #contar fechas fuera de marzo en datos limpio

Viajes excluidos por duración: 6
Viajes excluidos por velocidad: 11
Viajes excluidos por fecha: 1
Total de viajes excluidos: 18

Verificación:
Duraciones <= 0: 0
Velocidades > 60 mph: 0
Pickups fuera de marzo 2019: 0


---
---
* En el código del ejercicio 4.1 primero se crean las condiciones cond_duracion, cond_velocidad y cond_fecha para identificar viajes con duración menor o igual a 0, velocidad mayor a 60 mph y fechas fuera de marzo de 2019. Con `.sum()` cuento cuántos viajes se excluyen por cada condición. Después se usa `taxis_limpio = taxis[~(cond_duracion | cond_velocidad | cond_fecha)].copy() `para crear el dataset limpio, quedándo solo con los viajes que cumplen todas las condiciones. Luego se calculo el total de viajes excluidos con `len(taxis) - len(taxis_limpio)` y finalmente se volvieron a contar las condiciones en taxis_limpio para verificar que ya no haya duraciones menores o iguales a 0, velocidades mayores a 60 mph ni pickups fuera de marzo de 2019.
---
---

### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [45]:
# Escribe tu código aquí
horario = taxis_limpio['hora'].between(7, 9) #Viajes que inician entre las 7 y 9
habiles = taxis_limpio['fin_semana'] == False #Fin de semana
fin_s = taxis_limpio['fin_semana'] == True #Fin de semana

#Grupos
viajes_habiles = taxis_limpio[horario & habiles]
viajes_fin_semana = taxis_limpio[horario & fin_s]

#comparacion

comparacion = pd.DataFrame({
    'Grupo de dias': ['Días hábiles', 'Fin de semana'],
    'num_viajes': [
        len(viajes_habiles),
        len(viajes_fin_semana)
    ],
    'mediana_duracion': [
        viajes_habiles['duracion_min'].median(),
        viajes_fin_semana['duracion_min'].median()
    ],
    'mediana_velocidad': [
        viajes_habiles['velocidad_mph'].median(),
        viajes_fin_semana['velocidad_mph'].median()
    ]
})

comparacion = comparacion.round(1)

print(comparacion)




   Grupo de dias  num_viajes  mediana_duracion  mediana_velocidad
0   Días hábiles         687              12.0                9.1
1  Fin de semana         163               7.6               13.5


---
---
* En el código del ejercicio 4.2 primero uso `between(7, 9)` para seleccionar los viajes que empiezan entre las 7:00 y las 9:59. Después se agrupan a con `fin_semana == False` para los días hábiles y `fin_semana == True` para el fin de semana (previamente realizado en ejercicios anteriores). Con `horario & habiles` y `horario & fin_s` se filtran los viajes que cumplen ambas condiciones. Luego  se crea comparacion con` pd.DataFrame()` para mostrar, en cada grupo, el número de viajes usando `len()`, la mediana de duración con .`median()` y la mediana de velocidad también con `.median()`.
---
---

In [46]:
print('Antes de limpieza',len(taxis))
print('Despues de limpieza',len(taxis_limpio))
print('Total de excluidos', len(taxis)-len(taxis_limpio))


Antes de limpieza 6433
Despues de limpieza 6415
Total de excluidos 18


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**
Registro de limpieza realizada:
* **Conversión:** Se convierieron las columnas pickup y dropoff a formato de fecha y hora utilizando pd.to_datetime()y se verifico que las columnas tuvieran correctamente el tipo datetime, lo que permitió utilizar funciones como .dt.hour, .dt.date y .dt.dayofweek para extraer información de las fechas.

* **Zona horaria:** Para asignar la zona horaria a los registros de pickup se utilizo `dt.tz_localize('America/New_York', nonexistent='NaT')`. La zona America/New_York corresponde a la zona horaria local de Nueva York. El parámetro `nonexistent='NaT'` permite identificar las horas que no existen debido al cambio de horario de verano, convirtiéndolas en NaT en lugar de generar un error. Después con `.isna().sum()` se contaron cuántos registros quedaron como NaT (que fueron 0) y así verificar si existían problemas con horas no válidas.

* **Criterio de exclusión:** Para crear taxis_limpio se exluyeron los viajes que tuvieran una duracion_min menor o igual a 0, ya que una duración de cero no permite calcular correctamente la velocidad. También los viajes con una velocidad_mph mayor a 60 mph, porque se presentaron valores exagerados, como velocidades de hasta 4834 mph, que no son razonables para un viaje en taxi. Adicionalmente se mantuvieron únicamente los viajes cuyo pickup estuviera dentro de marzo de 2019 No se eliminaron automáticamente los viajes menores a 1 minuto, porque algunos pueden tener distancias y tarifas razonables por lo que vale la pena continuar analizando el conxtexto e interferencia de los mismos.

* **Resultado:** Antes de la limpieza había 6433 viajes. Después de aplicar los criterios quedaron 6415 viajes, por lo que se excluyeron 18 viajes en total. Las exclusiones fueron: 6 por duración, 11 por velocidad y 1 por fecha. Al verificar taxis_limpio, las condiciones se cumplen: no hay viajes con duración menor o igual a 0, no hay velocidades superiores a 60 mph y todos los pickups pertenecen a marzo de 2019.

* **Hallazgo de 4.2:** En el análisis del horario de 7:00 a 9:59 se observaron diferencias entre los días hábiles y el fin de semana tanto en el número de viajes como en las medianas de duración y velocidad. Esto permite identificar que el comportamiento de los viajes puede cambiar dependiendo de si se trata de un día laboral o de fin de semana. Además, el análisis por día de la semana mostró diferencias en el número de viajes, por lo que fue necesario calcular el promedio de viajes por fecha para hacer una comparación más justa entre los días.

* **Limitaciones:** El análisis está limitado a los viajes registrados durante marzo de 2019, por lo que los resultados no necesariamente representan otros meses o años. Además, el límite de 60 mph es un criterio establecido para detectar valores poco razonables, pero no garantiza que todos los viajes restantes sean completamente correctos. También pueden existir otros errores en los datos que no sean detectados por las condiciones utilizadas. Finalmente, las medianas permiten comparar los grupos, pero no permiten determinar por sí solas las causas de las diferencias observadas.



---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. Indicar el formato explícito en 1.1 fue mejor porque le dije a pandas exactamente cómo interpretar las fechas, evitando que intentara adivinarlas. Por ejemplo, si usamos un formato como '%Y-%m-%d %H:%M:%S', pandas sabe qué parte corresponde al año, mes, día y hora. La inferencia podría producir errores silenciosos si algunas fechas tuvieran formatos ambiguos, como 03/04/2019, ya que podría interpretar 03/04 como 3 de abril o 4 de marzo, dependiendo de cómo lo interprete.

2. Si pickup estuviera en UTC y dropoff en hora local de Nueva York, al restarlas directamente las duraciones quedarían mal, porque las dos horas no estarían en la misma zona horaria. Por ejemplo, el viaje podría parecer varias horas más largo o más corto de lo que realmente fue. Lo detectaría revisando las zonas horarias con `tz_localize()`, comparando las fechas de pickup y dropoff y revisando si aparecen duraciones o velocidades exageradas.

3. Si no limpiáramos las anomalías temporales, el modelo podría aprender de duraciones incorrectas o valores extremos. Por ejemplo, un viaje con una duración exageradamente alta podría hacer que el modelo piense que ciertos horarios o días producen viajes mucho más largos de lo normal. Esto puede generar predicciones menos precisas y hacer que el modelo aprenda relaciones que realmente son consecuencia de errores en los datos y no del comportamiento real de los viajes. Además, los valores extremos pueden afectar especialmente a modelos sensibles a los outliers.


---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.